redis/src/sentinel.rs

```diff
@@ -0,0 +1,795 @@
+    ///
+    fn try_all_sentinels<T: FromRedisValue>(&mut self, cmd: Cmd) -> RedisResult<T> {
+        let mut last_err = None;
+        for (connection_info, cached_connection) in self
+            .sentinels_connection_info
+            .iter()
+            .zip(self.connections_cache.iter_mut())
+        {
+            match try_single_sentinel(cmd.clone(), connection_info, cached_connection) {
+                Ok(result) => {
+                    return Ok(result);
+                }
+                Err(err) => {
+                    last_err = Some(err);
+                }
+            }
+        }
+
+        // We can unwrap here because we know there is at least one connection info.
+        // Maybe we should have a more specific error for this situation?
+        Err(last_err.expect("There should be at least one connection info"))
+    }
+
+    /// Get a list of all masters (using the command SENTINEL MASTERS) from the
+    /// sentinels (we'll return the result of the first sentinel who responds to that
+    /// command without errors).
+    fn get_sentinel_masters(&mut self) -> RedisResult<Vec<HashMap<String, String>>> {
+        self.try_all_sentinels(sentinel_masters_cmd())
+    }
+
+    fn get_sentinel_replicas(
+        &mut self,
+        service_name: &str,
+    ) -> RedisResult<Vec<HashMap<String, String>>> {
+        self.try_all_sentinels(sentinel_replicas_cmd(service_name))
+    }
+
+    fn find_master_address(
+        &mut self,
+        service_name: &str,
+        node_connection_info: &SentinelNodeConnectionInfo,
+    ) -> RedisResult<ConnectionInfo> {
+        let masters = self.get_sentinel_masters()?;
+        find_valid_master(masters, service_name, node_connection_info)
+    }
+
+    fn find_valid_replica_addresses(
+        &mut self,
+        service_name: &str,
+        node_connection_info: &SentinelNodeConnectionInfo,
+    ) -> RedisResult<Vec<ConnectionInfo>> {
+        let replicas = self.get_sentinel_replicas(service_name)?;
+        Ok(get_valid_replicas_addresses(replicas, node_connection_info))
+    }
+
+    /// Determines the masters address for the given name, and returns a client for that
+    /// master.
+    ///
+    /// To determine the master we ask the sentinels one by one until one of
+    /// then correctly responds to the SENTINEL MASTERS command, and then we check to
+    /// see if there is a master with the given name; if there is, we create a
+    /// client with that address (ip and port) and the extra information from the
+    /// `node_connection_info` parameter.
+    pub fn master_for(
+        &mut self,
+        service_name: &str,
+        node_connection_info: Option<&SentinelNodeConnectionInfo>,
+    ) -> RedisResult<Client> {
+        let connection_info =
+            self.find_master_address(service_name, node_connection_info.unwrap_or_default())?;
+        Client::open(connection_info)
+    }
+
+    /// Connects to the first available replica of the given master name.
```

Comment:
This doc is a bit vague - what is the "first available replica"? I would just say "a replica" or "a randomly chosen replica", and drop the second paragraph.

Reply:
Like I said above, I agree that it doesn't make sense, I've replaced it as you've suggested!